src/backend/distributed/executor/adaptive_executor.c

```diff
@@ -1013,12 +1010,30 @@ ExecuteTaskListExtended(ExecutionParams *executionParams)
 		executionParams->targetPoolSize = 1;
 	}
 
+	if (executionParams->localExecutionSupported == false)
+	{
+		/*
+		 * Later we may decide failing over to local execution if remote
+		 * execution fails for some reason. However, we shouldn't do that if
+		 * caller told us to avoid from local execution.
+		 *
+		 * SetLocalExecutionStatus(LOCAL_EXECUTION_DISABLED) already throws
+		 * an error if we try disabling local execution (LOCAL_EXECUTION_DISABLED)
+		 * after accessing any local placements (LOCAL_EXECUTION_REQUIRED).
+		 *
+		 * However, to give a better error, we first call
+		 * ErrorIfTransactionAccessedPlacementsLocally here.
+		 */
+		ErrorIfTransactionAccessedPlacementsLocally();
+		SetLocalExecutionStatus(LOCAL_EXECUTION_DISABLED);
+	}
```

Comment:
I didn't want to embed that thing into `CreateDistributedExecution`, as it's only supposed to create & return a struct.

Reply:
I think it would be slightly nicer to do it there to make sure we cover all possible cases.

Reply:
Or, what about pushing below parts down to `StartDistributedExecution()` ? I would rather expect `StartDistributedExecution()` to alter global state than `CreateDistributedExecution()`.

```c
	if (executionParams->localExecutionSupported == false)
	{
		/*
		 * Later we may decide failing over to local execution if remote
		 * execution fails for some reason. However, we shouldn't do that if
		 * caller told us to avoid from local execution.
		 *
		 * SetLocalExecutionStatus(LOCAL_EXECUTION_DISABLED) already throws
		 * an error if we try disabling local execution (LOCAL_EXECUTION_DISABLED)
		 * after accessing any local placements (LOCAL_EXECUTION_REQUIRED).
		 *
		 * However, to give a better error, we first call
		 * ErrorIfTransactionAccessedPlacementsLocally here.
		 */
		ErrorIfTransactionAccessedPlacementsLocally();
		SetLocalExecutionStatus(LOCAL_EXECUTION_DISABLED);
	}

	/*
	 * If current transaction accessed local placements and task list includes
	 * tasks that should be executed locally (accessing any of the local placements),
	 * then we should error out as it would cause inconsistencies across the
	 * remote connection and local execution.
	 */
	List *remoteTaskList = execution->remoteTaskList;
	if (GetCurrentLocalExecutionStatus() == LOCAL_EXECUTION_REQUIRED &&
		AnyTaskAccessesLocalNode(remoteTaskList))
	{
		ErrorIfTransactionAccessedPlacementsLocally();
	}


```

Reply:
works for me

Reply:
(nvm, this wouldn't work)

Reply:
(or it could, let me quickly try)